### ISSUE TITLE
Finding the number of classes

### ISSUE THREAD

**Issue by HelloDraemer1234** [mashijie1028/ProtoGCD#3]
Well done! However, I have a question regarding how you estimate the number of new classes. I could not find the definitions of the two scores mentioned in your paper within the provided code. Additionally, if it is convenient for you, could you please provide an example of how to estimate the number of new classes? Thank you.

**Comment by mashijie1028** [comment]
@HelloDraemer1234 
Hi! Thanks for your attention to our work. Here, we provide a code snippet of the prototype score (comprising accuracy and centroid scores) for class number estimation.
```python
def test_k(model, train_loader_k, args):

    model.eval()

    preds_labeled, targets_labeled = [], []
    #mask = np.array([])
    centroid_labeled = torch.zeros(args.num_labeled_classes, args.feat_dim).to(device)
    centroid_counts = [0] * args.num_labeled_classes
    # pseudo labelled centroids
    centroid_pseudo_whole_unlabeled = torch.zeros(args.num_labeled_classes + args.estimate_novel_k, args.feat_dim).to(device)   # NOTE! unlabeled
    centroid_pseudo_counts = [0] * (args.num_labeled_classes + args.estimate_novel_k)

    for batch_idx, (images, labels, _, mask_lab) in enumerate(tqdm(train_loader_k)):
        images = images.cuda(non_blocking=True)
        mask_lab = mask_lab[:, 0]
        labels, mask_lab = labels.cuda(non_blocking=True), mask_lab.cuda(non_blocking=True).bool()
        with torch.no_grad():
            feats, _, logits, prototypes = model(images)
            logits_old = logits[:, :args.num_labeled_classes]
            logits_whole = logits

            # mean centroid of labeled data
            feats = F.normalize(feats, dim=-1, p=2)
            feats_labeled = feats[mask_lab]
            feats_unlabeled = feats[~mask_lab]
            logits_old_labeled = logits_old[mask_lab]
            logits_whole_labeled = logits_whole[mask_lab]
            logits_whole_unlabeled = logits_whole[~mask_lab]
            labels_labeled = labels[mask_lab]
            for k, feature in enumerate(feats_labeled):
                centroid_labeled[labels_labeled[k].item()] += feature
                centroid_counts[labels_labeled[k].item()] += 1

            preds_labeled.append(logits_whole_labeled.argmax(1).cpu().numpy())   # NOTE! logits_whole rather than logits_old
            targets_labeled.append(labels_labeled.cpu().numpy())
            #mask = np.append(mask, np.array([True if x.item() in range(len(args.train_classes)) else False for x in label]))

            # pseudo_labels_whole = logits_whole.argmax(1)
            pseudo_labels_whole_unlabled = logits_whole_unlabeled.argmax(1)   # NOTE!!!
            for k, feature in enumerate(feats_unlabeled):
                centroid_pseudo_whole_unlabeled[pseudo_labels_whole_unlabled[k].item()] += feature
                centroid_pseudo_counts[pseudo_labels_whole_unlabled[k].item()] += 1

    for c in range(args.num_labeled_classes):
        centroid_labeled[c] /= centroid_counts[c]
    centroid_labeled = F.normalize(centroid_labeled, dim=-1, p=2.0)

    for c in range(args.num_labeled_classes + args.estimate_novel_k):
        centroid_pseudo_whole_unlabeled[c] /= centroid_pseudo_counts[c]
    centroid_pseudo_whole_unlabeled = F.normalize(centroid_pseudo_whole_unlabeled, dim=-1, p=2.0)

    # score1: prototype shift (similarity)
    centroid_pseudo_old_unlabeled = centroid_pseudo_whole_unlabeled[:args.num_labeled_classes]
    score_sim_avg = torch.trace(centroid_labeled @ centroid_pseudo_old_unlabeled.T).item()
    score_sim_avg /= (args.num_labeled_classes + 1e-6)
    score_sim_avg = (score_sim_avg + 1) / 2   # scale from [-1,1] to [0,1]
    score_sim_prod = torch.prod(torch.diagonal(centroid_labeled @ centroid_pseudo_old_unlabeled.T)).item()

    # score2: acc on labelled data
    preds_labeled = np.concatenate(preds_labeled)
    targets_labeled = np.concatenate(targets_labeled)
    preds_labeled = preds_labeled.astype(int)
    targets_labeled = targets_labeled.astype(int)

    #old_acc = old_cluster_acc(targets_labeled, preds_labeled)
    old_acc = np.sum(targets_labeled == preds_labeled) / len(targets_labeled)

    score_acc = old_acc

    prototype_score_avg = score_sim_avg * score_acc
    prototype_score_prod = score_sim_prod * score_acc

    return prototype_score_avg, prototype_score_prod, score_sim_avg, score_sim_prod, score_acc
```

**Comment by HelloDraemer1234** [comment]
Hi! Thank you so much for your prompt and helpful reply! That will help me a lot.
